### PR TITLE
Big commit that fixes NetCDF write failures that occurred with WriteO…

### PR DESCRIPTION
**Pull request by mdfischer**
…utputAllCells::write_data() when unused cells exist in the domain as a result of a non-rectangular spatial extent of active cells. Uses a mask to indicate which cells among the larger NetCDF domain are active (real) cells to be written to with real data, and which are to be written with a NetCDF FillValue.

### CODE REVIEW

**Comment by jameshiebert** [comment]
I think you need to rebase this branch onto master to resolve the conflicts.

File: get_global_param.c [hunk 1]
```diff
@@ -102,11 +102,41 @@ void ProgramState::initGrid(const std::vector<cell_info_struct>& cells) {
...
+  char            ErrStr[MAXSTRING];
+	int total_num_cells = global_param.gridNumLatDivisions * global_param.gridNumLonDivisions;
+	valid_cell_mask = new bool [total_num_cells];
+	float lat = global_param.gridStartLat;
+	float lon = global_param.gridStartLon;
+  int cellidx = 0;
+  int count = 0;
+
+	// Check all possible lat/lon pairs in the grid and check if each exists within valid_cell_coordinates set
+	while (lat <= global_param.gridEndLat){
```
Comment (jameshiebert) [inline get_global_param.c]: Use `for` loops here. You're iterating over a fixed number of things, and using a `for` loop will better communicate that to the maintainer of this code.

File: WriteOutputAllCells.c [hunk 2]
```diff
@@ -36,7 +36,8 @@ void WriteOutputAllCells::write_data(std::vector<out_data_struct*>& all_out_data
   std::vector<size_t> start4(start4Vals, start4Vals + 4), count4(count4Vals, count4Vals + 4);
```
Comment (jameshiebert) [inline WriteOutputAllCells.c]: This entire file has mixed (tabs _and_ spaces) whitespace, which is really obnoxious to read. Can you please fix this?

File: get_global_param.c [hunk 3]
```diff
@@ -102,11 +102,41 @@ void ProgramState::initGrid(const std::vector<cell_info_struct>& cells) {
...
+		while (lon <= global_param.gridEndLon){
+			count = valid_cell_coordinates.count(std::make_tuple(lat, lon));
+			if (count == 1)
+				valid_cell_mask[cellidx] = true;
+			else if (count == 0)
+				valid_cell_mask[cellidx] = false;
+			else {
```
Comment (jameshiebert) [inline get_global_param.c]: Will this actually find duplicate cells? (Hint: I don't think it will.) A [set](http://www.cplusplus.com/reference/set/set/), by definition has unique items, so the `count()` [can only be zero or one](http://www.cplusplus.com/reference/set/set/count/).

File: WriteOutputNetCDF.c [hunk 4]
```diff
@@ -282,7 +281,7 @@ void WriteOutputNetCDF::initializeFile(const ProgramState* state, const out_data
           data.putAtt("units", metaData.units.c_str());
           data.putAtt("standard_name", metaData.standardName.c_str());
           data.putAtt("cell_methods", metaData.cellMethods.c_str());
-          data.putAtt("_FillValue", ncFloat, 1e20f);
+          data.putAtt("_FillValue", ncFloat, NETCDF_FILL_VALUE);
```
Comment (jameshiebert) [inline WriteOutputNetCDF.c:284]: Good call, in eliminating random constants from the code.

File: get_global_param.c [hunk 5]
```diff
@@ -73,8 +73,8 @@ void ProgramState::initGrid(const std::vector<cell_info_struct>& cells) {
     global_param.gridStartLon = cells[0].soil_con.lng;
     global_param.gridStepLat = INT_MAX;
     global_param.gridStepLon = INT_MAX;
-    double largestLat = cells[0].soil_con.lat;
-    double largestLon = cells[0].soil_con.lng;
+    global_param.gridEndLat = cells[0].soil_con.lat;
```
Comment (jameshiebert) [inline get_global_param.c:76]: Why the change in variable names?

File: get_global_param.c [hunk 6]
```diff
@@ -102,11 +102,41 @@ void ProgramState::initGrid(const std::vector<cell_info_struct>& cells) {
...
+    global_param.gridNumLatDivisions = ((global_param.gridEndLat - global_param.gridStartLat) / global_param.gridStepLat) + 1;
+    global_param.gridNumLonDivisions = ((global_param.gridEndLon - global_param.gridStartLon) / global_param.gridStepLon) + 1;
   }
 }
 
+void ProgramState::initCellMask(const std::vector<cell_info_struct>& cells) {
+
+  char            ErrStr[MAXSTRING];
+	int total_num_cells = global_param.gridNumLatDivisions * global_param.gridNumLonDivisions;
+	valid_cell_mask = new bool [total_num_cells];
```
Comment (jameshiebert) [inline get_global_param.c]: Is there a good reason for the mask to be a simple array rather than being two dimensional? Ultimately, it's the same information, but it just is weird to have the mask disagree in structure with everything else about the grid. If it were 2d, then the iteration matches up with both of your loops, whereas as you've written it, it's awkward to only increment your index inside the inner loop.

**Comment by jameshiebert** [comment]
Overall, I don't think that this PR is fundamentally flawed or anything. It looks like a reasonable solution to the problem. Not to diminish them, but most of the changes for which I have asked are aesthetic, so it shouldn't be too bad to work this into shape.

The last thing that I'd ask for is to use better terminology than "valid/invalid" to describe whether a given grid cell is being modelled or not. If a cell is outside of the spatial domain being modelled, it's not really "invalid", it just doesn't exist as far as the software is concerned. Perhaps we could use "modelled vs. unmodelled" or "inside vs. outside". The latter assumes that all of the grid cells are contiguous within some area, which doesn't necessarily have to the case, so I'd probably prefer the former.

Could you adjust this language throughout the changes?

File: get_global_param.c (hunk 3 above)
Comment (mdfischer) [inline get_global_param.c]: Ah, you're right. My C++ book didn't mention this.  IMHO 'count' is a poor choice for a function name.  It should be something like 'is_present'

File: get_global_param.c (hunk 5 above)
Comment (mdfischer) [inline get_global_param.c:76]: Just to be more consistent with the naming already existing for global_parm struct members, as these have been added to that struct (i.e. they are no longer temporary local variables within ProgramState::initGrid) because they need to be used in the new ProgramState::initCellMask() function.

File: get_global_param.c (hunk 6 above)
Comment (mdfischer) [inline get_global_param.c]: It's easier in that 1) we don't have to create additional i & j indices to track the stepping through lat and lon (which are floats), and 2) it maps perfectly (and is read in lock step) with the 1D vardata array that is populated in the loop inside WriteOutputAllCells::write_data() (lines 61-73) for the big write across all cells that happens for each variable.

File: get_global_param.c (hunk 3 above)
Comment (jameshiebert) [inline get_global_param.c]: FWIW, you probably want to use `unordered_set` since the order doesn't matter, ordering of tuples isn't even defined, and it will be quite a bit faster.

Comment (mdfischer) [inline get_global_param.c]: Yeah, although this would necessitate writing a custom hash and equality operator, as there is no built-in hash for tuple set elements (as there are for the basic data types).  Since this function is executed only during initialization, I think the performance improvement we might get from using unordered_set, if any, probably isn't worth the extra work.

File: get_global_param.c (hunk 6 above)
Comment (jameshiebert) [inline get_global_param.c]: They're not in lock step anyways... `valid_cell_idx` and `cell_idx` [are incremented under different circumstances](https://github.com/pacificclimate/VIC/blob/handle-ignored-cells/WriteOutputAllCells.c#L65). I agree somewhat that having the the dimensionality agree between the two has advantages. But the mapping between a 1d cell array and a 2d grid is already loose and has some dubious assumptions.

For example, what happens if a user provides a cell array that _doesn't_ proceed from smallest to largest latitude and from smallest to largest longitude? Remember that (for better or for worse) cells are defined in a big, long user input list by lat/lon with no order implied or assumed. If the user chooses a different cell ordering structure (or even a random order), which there's nothing stopping them from doing, this output code won't get the positions correct.

Comment (mdfischer) [inline get_global_param.c]: The WriteOutputAllCells::write_data() call to NcVar.putVar assumes that cells are contiguous in a C-language way, i.e. we have to write across all cells (modeled or not modeled) for a given time step, and have no other lat/lon information about them in particular at this stage.  Thus we have to go with the assumption that cells are ordered by increasing lat and then increasing lon.  This requirement has been discussed earlier with regards to multithreading -- ordering of cells in memory is necessary when processing multiple cells in parallel so that their results (which may not arrive in order temporally) can be slotted into the correct position for gathering up and writing out via write_data.  The user is responsible for ordering the soil file cells accordingly.  Alternatively, we could write extra code to accommodate un/incorrectly ordered soil files, and do the sorting during initialization.
